File: Patterns/State.py
from abc import ABC, abstractmethod


# Create a State Design pattern: Define an abstract class (interface)
class TrafficState(ABC):
    @abstractmethod
    def switch(self, traffic_light: "TrafficLight"):
        pass


# Define a concrete state class for green light
class GreenLight(TrafficState):
    def switch(self, traffic_light: "TrafficLight"):
        print("Switching from green to yellow light.")
        traffic_light.state = YellowLight()


# Define a concrete state class for yellow light
class YellowLight(TrafficState):
    def switch(self, traffic_light: "TrafficLight"):
        print("Switching from yellow light to red light.")
        traffic_light.state = RedLight()


# Define a concrete state class for red light
class RedLight(TrafficState):
    def switch(self, traffic_light: "TrafficLight"):
        print("Switching from red light to green light.")
        traffic_light.state = GreenLight()


# Create a concrete Traffic Light class to switch between states
class TrafficLight:
    def __init__(self):
        # Initialize the traffic light to red light
        self._state = RedLight()

    # Defining getter and setter for state
    @property
    def state(self):
        return self._state

    @state.setter
    def state(self, state: TrafficState):
        self._state = state

    def switch(self):
        self.state.switch(self)


# Test the system
traffic_light = TrafficLight()
for index in range(6):
    traffic_light.switch()

File: Java/Patterns/State.java
interface TrafficState {
    void switchState(TrafficLight trafficLight);
}

class GreenLight implements TrafficState {
    @Override
    public void switchState(TrafficLight trafficLight) {
        System.out.println("Switching from green to yellow light.");
        trafficLight.setState(new YellowLight());
    }
}

class YellowLight implements TrafficState {
    @Override
    public void switchState(TrafficLight trafficLight) {
        System.out.println("Switching from yellow light to red light.");
        trafficLight.setState(new RedLight());
    }
}

class RedLight implements TrafficState {
    @Override
    public void switchState(TrafficLight trafficLight) {
        System.out.println("Switching from red light to green light.");
        trafficLight.setState(new GreenLight());
    }
}

class TrafficLight {
    private TrafficState state;

    // Initialize the traffic light with the Red Light state
    public TrafficLight() {
        this.state = new RedLight();
    }

    // Getter for the current state
    public TrafficState getState() {
        return state;
    }

    // Setter for the current state
    public void setState(TrafficState state) {
        this.state = state;
    }

    // Method to trigger state transition
    public void switchState() {
        state.switchState(this);
    }
}

// Test the system
public class State {
    public static void main(String[] args) {
        TrafficLight trafficLight = new TrafficLight();

        // Switch traffic light state 6 times
        for (int i = 0; i < 6; i++) {
            trafficLight.switchState();
        }
    }
}
